### PR TITLE
Fix replay staging buffer binding for when the capture/replay devices are not the same.

### PR DESCRIPTION
**Pull request by jacobv-nvidia**
While testing the replay of trimmed capture files on devices other than the original capture device, I encountered an issue in the rebind allocator where allocations would fail, causing the replay to abort.

The initial loading allocations would fail, and, when running the replay with the `--validate` option, I would receive messages such as the following:
```
VUID-vkMapMemory-memory-00682(ERROR / SPEC): msgNum: -330527817 - Validation Error: [ VUID-vkMapMemory-memory-00682 ] Object 0: handle = 0x4b07580000000ca2, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0xec4c8bb7 | Mapping Memory without VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT set: VkDeviceMemory 0x4b07580000000ca2[]. The Vulkan spec states: memory must have been created with a memory type that reports VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT (https://vulkan.lunarg.com/doc/view/1.3.239.0/windows/1.3-extensions/vkspec.html#VUID-vkMapMemory-memory-00682)
    Objects: 1
	[0] 0x4b07580000000ca2, type: 8, name: NULL
```

The issue seems to be in the rebind allocator's "Direct"` binding functions, which are supposed to be performed without memory translation for the purposes of, for example, replay staging buffers. However, these functions were using the capture device's memory types instead of the replay device's.

This patch should fix the issue.

### CODE REVIEW

**Review by andrew-lunarg — COMMENTED** [review]
Looks like a very useful fix, so thanks for the PR!

One point, It might be cleaner to extend `BindBufferMemory` and `BindImageMemory` with the extra bool parameter and give it a default value of true. Then the two new functions with slightly vague names wouldn't be needed.

**Comment by jacobv-nvidia** [comment]
Yes, I was on the fence about that. I was trying to avoid cluttering the parent interface of the more general class `VulkanResourceAllocator`. Do you think it's worth it?

**Review by locke-lunarg — APPROVED** [review]

**Review by bradgrantham-lunarg — CHANGES_REQUESTED** [review]
Thank you for this PR!  I have proposed a small change to the calling sequence in a comment.

**Comment by bradgrantham-lunarg** [comment]
I tried to take a trimmed capture of vkcube from llvmpipe and replay it on NVIDIA with `--validate` but got no validation error for vkMapMemory from SDK 1.3.246's binaries.@jacobv-nvidia , what's a good test case for this?

File: framework/decode/vulkan_rebind_allocator.cpp [hunk 1]
```diff
@@ -409,7 +410,9 @@ VkResult VulkanRebindAllocator::BindBufferMemory(VkBuffer               buffer,
         create_info.flags = 0;
         create_info.usage = GetBufferMemoryUsage(
             resource_alloc_info->usage,
-            capture_memory_properties_.memoryTypes[memory_alloc_info->original_index].propertyFlags,
+            is_direct_allocation
```
[the remark this answers is not in the document]
Comment (jacobv-nvidia) [reply framework/decode/vulkan_rebind_allocator.cpp:413]: Sure, go ahead. I agree, your new change seems like a much less hacky implementation of the fix.

**Comment by jacobv-nvidia** [comment]
> I tried to take a trimmed capture of vkcube from llvmpipe and replay it on NVIDIA with `--validate` but got no validation error for vkMapMemory from SDK 1.3.246's binaries.@jacobv-nvidia , what's a good test case for this?

I have just successfully tested it with vkcube, so that should work as a good test case, but you do need to:
1. use the vkcube option `--use_staging`
2. avoid including the first frame as part of the trim.

For example: `python gfxrecon-capture-vulkan.py -o cube.gfxr --capture-frames 2-16 vkcube --use_staging`

I haven't really tested on llvmpipe, so can't speak too much on that part. The issue was discovered and tested using an NVIDIA trace on an AMD card, but I suppose it's possible that the relevant memory type indices just happen to match up between llvmpipe and NVIDIA.

**Comment by bradgrantham-lunarg** [comment]
Merged as #1173.

We encountered this this week in our internal CI and this solved our problem so I prioritized it.

Thank you for finding this @jacobv-nvidia !

**Comment by jacobv-nvidia** [comment]
@bradgrantham-lunarg Apologies for the delay in responding, I thought I had replied to your earlier comment on your changes. Anyway, I appreciate the merge.

**Pull request closed without merging**